chain-signatures/node/src/storage/request_queue.rs

```diff
@@ -0,0 +1,52 @@
+//! A FIFO queue for signature requests between many chain indexers (producers)
+//! and one MPC node (consumer).
+//!
+//! Under the "sign_request_queue" key on redis, we store signature requests
+//! that have been validated by a chain indexer. The MPC node is picking them up
+//! as the sole consumer on the queue.
+//!
+//! The redis type is a list, therefore, use `lpush` and `brpop` to modify the
+//! queue.
+
+use crate::protocol::SignRequest;
+use redis::AsyncCommands;
+
+const QUEUE_KEY: &str = "sign_request_queue";
+
+/// Take a validated sign request from a chain and hand it off to the MPC node
+/// for processing.
+pub async fn push_sign_request(
+    redis_pool: &deadpool_redis::Pool,
+    request: SignRequest,
+) -> anyhow::Result<()> {
+    let mut request_cbor = Vec::new();
+    ciborium::ser::into_writer(&request, &mut request_cbor)?;
+
+    let mut conn = redis_pool.get().await?;
+    // Using a redis list as a FIFO queue, pushing on the left and popping on the right.
+    let _len: usize = conn.lpush(QUEUE_KEY, request_cbor).await?;
+
+    Ok(())
+}
+
+/// Take a signature request from the head of the FIFO queue.
+///
+/// Blocks until a request is available, asynchronously waiting for a
+/// notification by redis.
+pub async fn pop_sign_request(redis_pool: &deadpool_redis::Pool) -> anyhow::Result<SignRequest> {
+    let mut conn = redis_pool.get().await?;
+    // Using a redis list as a FIFO queue, pushing on the left and popping on the right.
+    // timeout = 0.0 will block indefinitely
+    let maybe_request_cbor: Option<Vec<u8>> = conn.brpop(QUEUE_KEY, 0.0).await?;
```

Comment:
Am I right that in practice we should not expect it to be infinite? Communication with Redis is done over the wire, if there are no signature requests for ~1H, it will fail.

Reply:
This isn't a timeout for the network request but rather on the redis side.

It will block in case of an empty queue, until a request is available to be popped. If there is none for 1h, the task does nothing for 1h. It won't ever fail, unless the connection to redis is lost.

Reply:
hmm, so this blocks the sign queue entirely then right? In that case, we'd need to split sign queue itself out to be its own separate task, or at least a task that purely just pops the requests and organizes the requests.

I'm already making the signature manager a separate task so not a huge issue to do this.

Reply:
Yeah, `SignQueue::organize` should run in its own task, if we block here. Or, we can also use non-blocking calls to the redis queue if you prefer that.